### PR TITLE
Move some code from Asmgen to the middle end directory etc.

### PR DESCRIPTION
**Pull request by mshinwell**
@chambart This is the followup to #2281.  It moves some code that, morally speaking, belongs in the middle end but has thus far lived in `Asmgen`.  The interfaces for Closure and Flambda are harmonised and various typos (e.g. "clambda" instead of "Closure") have been fixed.  In the spirit of not reaching back into the middle end from the backend, `Un_anf` is not called from `Cmmgen` any more, but always invoked from `Flambda_middle_end`.  We also pass constants used for instrumentation directly out of `Flambda_to_clambda` so they flow to `Cmmgen` without going through `Compilenv` (this will be important for a later patch).

This patch includes #2280 (just merged into trunk), #2281 and #2284.  For the moment just read changeset 4a3b0a8.

### CODE REVIEW

**Comment by mshinwell** [comment]
@chambart @lthls Could one of you please review and approve this if you are happy?  I've fixed the mistake in the comment that @lthls told me about.  This is looking like quite a nice patch now, it's simplified various parts in the backend and made the middle-end interfaces consistent.

Please double-check that `Un_anf` is being run as many times as it used to be -- the invocation point has changed.

**Comment by mshinwell** [comment]
P.S. For reviewing `closure_middle_end.ml` and `flambda_middle_end.ml` I suggest you `patdiff` against the old `asmgen.ml`.  There have been a few changes but those blocks are mainly moved verbatim IIRC.

File: asmcomp/asmpackager.ml [hunk 1]
```diff
@@ -99,26 +99,44 @@ let make_package_object ~ppf_dump members targetobj targetname coercion
...
+        { Lambda.
+          code;
+          main_module_block_size;
+          module_ident;
+          required_globals;
+        }
       in
-      Asmgen.compile_implementation_flambda
-        prefixname ~backend ~required_globals:Ident.Set.empty ~ppf_dump flam;
+      Asmgen.compile_implementation ~backend
```
Comment (chambart) [inline asmcomp/asmpackager.ml]: That call could be factorised out.
```
let program, middle_end =
  if Config.flambda then
   ...
in
Asmgen.compile_implementation ...
```
I'm not certain that this is better.

File: asmcomp/asmgen.ml [hunk 2]
```diff
@@ -167,109 +131,50 @@ let compile_unit _output_prefix asm_filename keep_asm
...
 
-let set_export_info (ulambda, prealloc, structured_constants, export) =
-  Compilenv.set_export_info export;
-  (ulambda, prealloc, structured_constants)
-
 let end_gen_implementation ?toplevel ~ppf_dump
-    (clambda:clambda_and_constants) =
+    (clambda : Clambda.with_constants) =
   Emit.begin_assembly ();
+  Cmmgen.reset ();
```
Comment (chambart) [inline asmcomp/asmgen.ml]: Why was that required ? Ins't the one at the beginning of compile_fundecl sufficient ?

By the way, by quickly looking at it, no reset would be required if `Cmmgen_state.constants` and `data_items` where cleaning their state.

File: middle_end/flambda/flambda_middle_end.ml [hunk 3]
```diff
@@ -198,3 +195,51 @@ let middle_end ~ppf_dump ~prefixname ~backend
...
+  let export = Build_export_info.build_transient ~backend program in
+  let clambda, preallocated_blocks, constants =
+    Profile.record_call "backend" (fun () ->
+      (program, export)
+      |> Flambda_to_clambda.convert ~ppf_dump
+      |> flambda_raw_clambda_dump_if ppf_dump
+      |> (fun { Flambda_to_clambda. expr; preallocated_blocks;
+                structured_constants; exported; } ->
+           Compilenv.set_export_info exported;
+           let clambda = Un_anf.apply ~ppf_dump expr in
```
Comment (chambart) [inline middle_end/flambda/flambda_middle_end.ml]: This code has a linkage_name, it could be printed. It doesn't really matter.

File: middle_end/flambda/flambda_to_clambda.ml [hunk 4]
```diff
@@ -70,22 +73,29 @@ let is_function_constant t closure_id =
...
   if not !Clflags.clambda_checks then ulam
   else
     let desc =
       Primitive.simple ~name:"caml_check_value_is_closure"
         ~arity:2 ~alloc:false
     in
     let str = Format.asprintf "%a" Flambda.print_named named in
-    let str_const =
+    let sym =
       Compilenv.new_structured_constant (Uconst_string str) ~shared:true
```
Comment (chambart) [inline middle_end/flambda/flambda_to_clambda.ml]: This doesn't just create a symbol name, it also records the structured constant. This should probably be replaced by 'new_const_symbol'

Or at least, flambda should use `clear_structured_constants` if this is used.

File: middle_end/flambda/flambda_to_clambda.ml [hunk 5]
```diff
@@ -97,11 +107,18 @@ let check_field ulam pos named_opt : Clambda.ulambda =
       | None -> "<none>"
       | Some named -> Format.asprintf "%a" Flambda.print_named named
     in
-    let str_const =
+    let sym =
       Compilenv.new_structured_constant (Uconst_string str) ~shared:true
```
Comment (chambart) [inline middle_end/flambda/flambda_to_clambda.ml]: Same thing

**Review by chambart — APPROVED** [review]
This looks good. The comments are minor things except the thing about new_structured_constants that should be fixed.

The result is quite cleaner, this is a good change.

File: asmcomp/asmpackager.ml (hunk 1 above)
Comment (mshinwell) [reply asmcomp/asmpackager.ml]: It does seem to be better.

File: asmcomp/asmgen.ml (hunk 2 above)
Comment (mshinwell) [reply asmcomp/asmgen.ml]: I've fixed things so `reset` is not needed.

File: middle_end/flambda/flambda_middle_end.ml (hunk 3 above)
Comment (mshinwell) [reply middle_end/flambda/flambda_middle_end.ml]: This should be fixed now.

File: middle_end/flambda/flambda_to_clambda.ml (hunk 4 above)
Comment (mshinwell) [reply middle_end/flambda/flambda_to_clambda.ml]: Fixed

File: middle_end/flambda/flambda_to_clambda.ml (hunk 5 above)
Comment (mshinwell) [reply middle_end/flambda/flambda_to_clambda.ml]: Fixed

**Comment by mshinwell** [comment]
@chambart All comments addressed; if you are happy and this passes CI please squash and merge.

**Comment by chambart** [comment]
This looks nice. merged.
